spark/datadog_checks/spark/spark.py

```diff
@@ -261,9 +263,36 @@ def _get_running_apps(self, instance):
             running_apps = self._yarn_init(master_address, tags)
             return self._get_spark_app_ids(running_apps, tags)
 
+        elif cluster_mode == SPARK_KUBERNETES_MODE:
+            return self._kubernetes_init(master_address, tags)
+
         else:
             raise Exception('Invalid setting for %s. Received %s.' % (SPARK_CLUSTER_MODE, cluster_mode))
 
+    def _kubernetes_init(self, spark_master_address, tags):
+        """
+        Return a dictionary of {app_id: (app_name, tracking_url)} for the running Spark applications
+        """
+        running_apps = {}
+        metrics_json = self._rest_request_to_json(
+            spark_master_address, SPARK_APPS_PATH, SPARK_KUBERNETES_SERVICE_CHECK, tags
```

Comment:
Should we reach Kubernetes instead of `SPARK_APPS_PATH` to retrieve the running_apps ?

In the same way we are calling yarn and mesos: https://github.com/DataDog/integrations-core/blob/master/spark/datadog_checks/spark/data/conf.yaml.example#L38-L47

Adding a section here for Kubernetes could help making this more clear: https://github.com/DataDog/integrations-core/blob/31ce02e7b02034050cb320e17eacdeeed4bd0256/spark/datadog_checks/spark/data/conf.yaml.example#L38-L47

Reply:
Hi! 
it is a bit different In case of K8s compare to Mesos and Yarn, it only uses k8s scheduler to bring up spark tasks, but k8s native api won't expose any spark specific API endpoint(Master web UI/API) to get list of tasks as in yarn or mesos, as you can see in http://spark.apache.org/docs/latest/running-on-kubernetes.html#future-work the "Spark Application Management" is not yet implemented, that why the spark_url should be the spark driver pod ip(SPARK_APPS_PATH) which is spinned up by k8s.
I can add some a section for k8s in the conf.yaml to make it clearer

Reply:
Hi @AlexandreYang ,
Thank you for your reply, I understand your concern, but I am not quite sure if it should be the way to be implemented in the k8s context.
Our spark driver pod is been brought up like this (remark the annotations): 
```yaml
$ % kubectl get pods k8s-spark-example-driver -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    ad.datadoghq.com/spark-kubernetes-driver.check_names: '["spark"]'
    ad.datadoghq.com/spark-kubernetes-driver.init_configs: '[{}]'
    ad.datadoghq.com/spark-kubernetes-driver.instances: '[{"spark_url":"http://%%host%%:4040","spark_cluster_mode":"spark_kubernetes_mode","cluster_name":"k8s-spark"}]'
    cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
  creationTimestamp: 2019-10-31T10:43:47Z
  labels:
    spark-app-selector: spark-5565b6cb35d3415aab210be6bd616e6b
    spark-role: driver
  name: k8s-spark-example-driver
  namespace: default
```
as you can see, we rely on the datadog k8s auto-discovery to get to the spark-ui url from the pod annotations `ad.datadoghq.com/spark-kubernetes-driver.instances : "spark_url":"http://%%host%%:4040"`, so the Step 1 as you mentioned is been accomplished by the datadog k8s auto-discovery functionality. if I put on each of my spark application driver's pod with the annotation : `ad.datadoghq.com/spark-kubernetes-driver.instances: '[{"spark_url":"http:/[KUBERNETES_API_ENDPOINT]",` then there's no point for the pod auto-discovery functionality.
Your suggestion is valid if we want to integrate the spark checks in the defaut kubenetes cluster checks, so the datadog agent daemonset will scan all the pods in the cluster to identify spark driver pods and automatically make spark checks on each of them.

Please let me know if it makes sense to you. you can check our example app of spark on kubernetes [https://github.com/flix-tech/k8s-spark-example](here), it will spin up a spark app on a local minikube with datadog autodiscovery annotations on the spark driver pod.
Kind regards

Reply:
Hi @mrmuggymuggy, thanks for the explanation, I understand better the use case.

If we want to monitor one single spark driver pod, then that would mean, this implementation is not specific to Kubernetes, but can be used in any case we want to monitor a single spark driver. Is that correct ?

If so, something like `spark_driver_mode` may be more appropriate than `spark_kubernetes_mode`. The config would be `spark_cluster_mode: "spark_driver_mode"`. One concern is that, the term cluster in `spark_cluster_mode` is not appropriate for this case, we are only monitor one driver.

Reply:
Hi @AlexandreYang , thanks for your reply, you are absolutely right, we only want to monitor the spark metrics from the specific spark driver, so it is not specific to k8s, `spark_kubernetes_mode` is misleading here, I can change it to `spark_driver_mode`. For the `spark_cluster_mode` naming concerns, we can maybe rename it to `spark_mode`? but I am afraid it will break many existing apps monitorings. please let me know if you have any idea to resolve the `cluster` term problem. As we move our spark apps from aws EMR to k8s. it will help us a lot to have proper monitoring. for the moment, we just patched the current spark checks to the datadog agent docker image deployed on as daemon-sets to have spark monitorings.  
kind regards

Reply:
@mrmuggymuggy I think it's reasonable to just change "Kubernetes" to "Driver", e.g. `spark_kubernetes_mode` into `spark_driver_mode` and all other occurrences.

Once this change is made, the PR should be ok to be merged (might need another review, but things look fine overall to me).

Reply:
Hi @AlexandreYang I have updated my PR and tests to remove kubernetes references and replace with `driver`. kind regards

Reply:
@mrmuggymuggy Looks good to me. Seems there are some conflicts with master, need a rebase.

Reply:
hey @AlexandreYang , I have rebased to master and squashed my commits. please review.